zcash_extensions/src/transparent/demo.rs

```diff
@@ -0,0 +1,734 @@
+        }
+    }
+}
+
+fn builder_hashes(preimage_1: &[u8; 32], preimage_2: &[u8; 32]) -> ([u8; 32], [u8; 32]) {
+    let hash_2 = {
+        let mut hash = [0; 32];
+        hash.copy_from_slice(Params::new().hash_length(32).hash(preimage_2).as_bytes());
+        hash
+    };
+
+    let hash_1 = {
+        let mut hash = [0; 32];
+        hash.copy_from_slice(
+            Params::new()
+                .hash_length(32)
+                .to_state()
+                .update(preimage_1)
+                .update(&hash_2)
+                .finalize()
+                .as_bytes(),
+        );
+        hash
+    };
+
+    (hash_1, hash_2)
+}
+
+pub struct DemoBuilder<B> {
+    pub txn_builder: B,
+    pub extension_id: u32,
+}
+
+#[derive(Debug)]
+pub enum DemoBuildError<E> {
+    BaseBuilderError(E),
+    ExpectedOpen,
+    ExpectedClose,
+    PrevoutParseFailure(Error),
+    TransferMismatch {
+        expected: [u8; 32],
+        actual: [u8; 32],
+    },
+    CloseMismatch {
+        expected: [u8; 32],
+        actual: [u8; 32],
+    },
+}
+
+impl<'a, B: ExtensionTxBuilder<'a>> DemoBuilder<&mut B> {
+    pub fn demo_open(
+        &mut self,
+        value: Amount,
+        preimage_1: [u8; 32],
+        preimage_2: [u8; 32],
+    ) -> Result<(), DemoBuildError<B::BuildError>> {
+        let (hash_1, _) = builder_hashes(&preimage_1, &preimage_2);
+
+        // Call through to the generic builder.
+        self.txn_builder
+            .add_tze_output(self.extension_id, value, &Precondition::open(hash_1))
+            .map_err(DemoBuildError::BaseBuilderError)
+    }
+
+    pub fn demo_transfer_to_close(
+        &mut self,
+        prevout: (OutPoint, TzeOut),
+        transfer_amount: Amount,
+        preimage_1: [u8; 32],
+        preimage_2: [u8; 32],
+    ) -> Result<(), DemoBuildError<B::BuildError>> {
+        let (hash_1, hash_2) = builder_hashes(&preimage_1, &preimage_2);
+
+        // eagerly validate the relationship between prevout.1 and preimage_1
+        match Precondition::from_payload(
+            prevout.1.precondition.mode,
+            &prevout.1.precondition.payload,
+        ) {
+            Ok(Precondition::Open(hash)) => {
+                if hash.0 != hash_1 {
+                    Err(DemoBuildError::TransferMismatch {
+                        expected: hash.0,
+                        actual: hash_1,
+                    })?
+                }
+            }
+            Ok(Precondition::Close(_)) => Err(DemoBuildError::ExpectedOpen)?,
+            Err(parse_failure) => Err(DemoBuildError::PrevoutParseFailure(parse_failure))?,
+        }
+
+        self.txn_builder
+            .add_tze_input(self.extension_id, prevout, move |_| {
+                Ok(Witness::open(preimage_1))
+            })
+            .map_err(DemoBuildError::BaseBuilderError)?;
+
+        self.txn_builder
+            .add_tze_output(
+                self.extension_id,
+                transfer_amount, // can this be > prevout.1.value?
```

Comment:
The demo TZE consensus rules allow for multiple (same-extension) TZE inputs as long as there is only one TZE output. So I think the answer is that `transfer_amount` could be larger than `prevout.1.value` if there were several demo open preconditions that used the same `preimage_1` and `hash_2` values - in effect, causing this close transaction to be a merging transaction.

Reply:
:heavy_check_mark: